SemanticChunker : Feature Addition ("Semantic Splitting with gradient")

rrajp opened the pull request:
```SemanticChunker``` currently provide three methods to split the texts semantically:
- percentile
- standard_deviation
- interquartile

I propose new method ```gradient```. In this method, the gradient of distance is used to split chunks along with the percentile method (technically) . This method is useful when chunks are highly correlated with each other or specific to a domain e.g. legal or medical. The idea is to apply anomaly detection on gradient array so that the distribution become wider and easy to identify boundaries in highly semantic data.
I have tested this merge on a set of 10 domain specific documents (mostly legal).

Details : 
    - **Issue:** Improvement
    - **Dependencies:** NA
    - **Twitter handle:** [x.com/prajapat_ravi](https://x.com/prajapat_ravi)


@hwchase17

Commit:
Feature Addition in semantic text_splitter method

In this method, the gradient of distance is used to split chunks along with the percentile method. This method is useful when chunks are highly correlated with each other or specific to a domain e.g. legal or medical. The idea is to apply anomaly detection on gradient array so that the distribution become wider and easy to identify boundaries in highly semantic data.

isahers1 commented on docs/docs/how_to/semantic-chunker.ipynb:
```diff
@@ -297,13 +297,85 @@
+   "metadata": {},
+   "id": "a407cd57f02a0db4"
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "outputs": [],
+   "source": [],
+   "metadata": {},
+   "id": "3b45b2b24672590f"
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "outputs": [],
+   "source": [],
+   "metadata": {
+    "collapsed": false
+   },
+   "id": "256b9e486f54663d"
```
Can you delete the empty code boxes at the bottom of the doc

rrajp replied:
Sorry, Missed that.
Deleted!

isahers1 commented on libs/experimental/langchain_experimental/text_splitter.py:
```diff
@@ -127,23 +128,30 @@ def __init__(
+            return np.mean(distances) + self.breakpoint_threshold_amount * iqr, distances
+        elif self.breakpoint_threshold_type == "gradient":
+            # Calculate the threshold based on the distribution of gradient of distance array.
+            distance_gradient = np.gradient(distances, list(range(0, len(distances))))
```
No need for the list(range... part

rrajp replied:
Yeah, Fixed that. Thanks for suggestion

isahers1 commented:
@baskaryan two small nits, and even though he changed function signature, it only gets called in class and he updated that call

isahers1 committed:
fmt

isahers1 committed:
fmt

Commit:
Feature Addition in semantic text_splitter method

Commit : Updated the code as suggested by @isahers1

In this method, the gradient of distance is used to split chunks along with the percentile method. This method is useful when chunks are highly correlated with each other or specific to a domain e.g. legal or medical. The idea is to apply anomaly detection on gradient array so that the distribution become wider and easy to identify boundaries in highly semantic data.

Commit:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	libs/experimental/langchain_experimental/text_splitter.py

rrajp committed:
Merge branch 'master' into master

rrajp committed:
Merge branch 'master' into master

rrajp commented:
@baskaryan Please review this. Let me know if there is any conflict or issue.

isahers1 committed:
fmt
